mlir/include/mlir/Dialect/Quant/IR/QuantTypes.h

```diff
@@ -224,6 +227,93 @@ class AnyQuantizedType
                    int64_t storageTypeMax);
 };
 
+/// Represents block floating point quantization where multiple elements share
+/// data along a particular axis (e.g. BFP16). The concrete block format
+/// determines the implied storage characteristics and is not exposed in the IR.
+/// This class is experimental and may be subject to change.
+/// Design decisions:
+///  - The base class requires an integral storage type. For
+///  block-quantized/packed types, the required storage with depends on the
+///  number of elements. For example, a single BFP16 element requires 16 bits to
+///  be represented, but a block of 8 BFP16 elements can be packed into 9 bits
+///  per element on average (72 bits total). The storage type for
+///  BlockFloatQuantizedType is the "packed" type, so for BFP16 i9.
+///  -- As accessing properties like min/max storage values and integral width
+///  depend on the block size, these methods are overridden to return errors.
+///  - The expressed type is not stored yet, this may change if there is a use
+///  for it.
+///  - The axis is signed to match MLIR convention, but enforced to be
+///  non-negative.
+class BlockFloatQuantizedType
+    : public Type::TypeBase<BlockFloatQuantizedType, QuantizedType,
+                            detail::BlockFloatQuantizedTypeStorage> {
+public:
+  using Base::Base;
+  using Base::getChecked;
+
+  static constexpr StringLiteral name = "quant.block_float";
+
+  // MX6 refers to the MicoExponent format, not to the OCP MicroScaling format
+  // with the same name.
+  enum class BlockMode : uint32_t { BFP16 = 0, MX6 = 1, MAX_VALUE = MX6 };
+
+  static std::optional<BlockMode> parseBlockMode(StringRef name);
+  static StringRef getBlockModeName(BlockMode blockMode);
+
+  static BlockFloatQuantizedType get(MLIRContext *ctx, BlockMode blockMode,
+                                     int32_t axis);
+  static BlockFloatQuantizedType
+  getChecked(function_ref<InFlightDiagnostic()> emitError, MLIRContext *ctx,
+             BlockMode blockMode, int32_t axis);
+
+  static LogicalResult
+  verifyInvariants(function_ref<InFlightDiagnostic()> emitError,
+                   uint32_t blockModeRaw, int32_t axis, unsigned flags,
+                   Type storageType, Type expressedType, int64_t storageTypeMin,
+                   int64_t storageTypeMax);
+
+  Type getStorageType() const {
+    assert(false &&
+           "BlockFloatQuantizedType does not have a direct storage type");
```

Comment:
NIT: Is there no other way to use something else instead of assert?
llvm_unreachable("BlockFloatQuantizedType does not have a direct storage type");